server/src/main/java/org/opensearch/action/ActionModule.java

```diff
@@ -553,6 +560,33 @@ public ActionModule(
         restController = new RestController(headers, restWrapper, nodeClient, circuitBreakerService, usageService, identityService);
     }
 
+    private Set<String> getAllRegisteredActionNames() {
```

Comment:
Suggested the approach with DynamicRestRouteRegistry (so we won't need to ActionModule around as well), what do you think? https://github.com/opensearch-project/OpenSearch/pull/6870/files#r1196825011

Reply:
One of the tests I had written was to verify that one of the names of the routes does not collide with a native action. i.e. if an extension tried to name a route `cluster:monitor/health` it would fail because that matches an action. I think action module would need to be around to get a set of transport action names.

Reply:
The overarching problem we are trying to solve with extensions is how to permission a user to use routes that have been registered by an extension. REST Requests are forwarded to an extension in [RestSendToExtensionAction](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java) so ideally REST Requests are blocked before they are forwarded to an extension if a user is not granted access to use that route.

With the Security plugin, its typically transport action names that are put into a role definition. With extensions, it won't hit the SecurityFilter of the Security plugin where authz checks are being performed so an additional authz layer is being placed into the SecurityRestFilter for requests bound for extensions. The goal of this PR was to be able to map those routes to a name that resembles a transport action name and grant access to the endpoint solely based on the name of the route.

Assuming that these permissions would also live in the `cluster_permissions` section of a role then they shouldn't collide with other transport action names or else there would be no way of differentiating them. I supposed its possible to add an new permissions section to a role. Perhaps `route_permissions` and place the names of these routes in that section, but I think @peternied had some hesitancy to add a new section on a role.

Reply:
Thanks @cwperks , I think it is clear now

Reply:
Hi @reta and @owaiskazi19 , I just pushed a change to introduce `DynamicRouteRegistry` to try to formalize the concept more. I was able to test most scenarios and tried adding a unit test that would create an action module with an action plugin that registers a route and verify no collision with plugin action names. The registry uses the data structure

```
Map<RestHandler.Route, RestSendToExtensionAction> registry
```

as the registry of dynamic routes and I'm hoping this may be useful going forward and perhaps provide a way in the future to eliminate the use of streams in places like this in the SDK: https://github.com/opensearch-project/opensearch-sdk-java/blob/main/src/main/java/org/opensearch/sdk/rest/BaseExtensionRestHandler.java#L88-L119